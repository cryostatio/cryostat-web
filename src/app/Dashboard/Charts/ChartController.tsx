/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { ApiService } from '@app/Shared/Services/Api.service';
import { NotificationCategory, NotificationChannel } from '@app/Shared/Services/NotificationChannel.service';
import { SettingsService } from '@app/Shared/Services/Settings.service';
import { NO_TARGET, Target, TargetService } from '@app/Shared/Services/Target.service';
import {
  BehaviorSubject,
  catchError,
  concatMap,
  distinctUntilChanged,
  finalize,
  first,
  map,
  merge,
  Observable,
  of,
  pairwise,
  ReplaySubject,
  Subscription,
  switchMap,
  tap,
  throttleTime,
} from 'rxjs';

export const RECORDING_NAME = 'dashboard_metrics';

export enum ControllerState {
  UNKNOWN = 0,
  NO_DATA = 1,
  READY = 2,
}

export class ChartController {
  private readonly _state$ = new BehaviorSubject<ControllerState>(ControllerState.UNKNOWN);
  private readonly _refCount$ = new BehaviorSubject<number>(0);
  private readonly _updates$ = new ReplaySubject<void>(1);
  private readonly _lazy: Subscription;
  private _attach: Subscription | undefined;

  constructor(
    private readonly _api: ApiService,
    private readonly _target: TargetService,
    private readonly _notifications: NotificationChannel,
    private readonly _settings: SettingsService
  ) {
    this._lazy = this._refCount$
      .pipe(
        map((v) => v > 0),
        pairwise()
      )
      .subscribe(([prev, curr]) => {
        if (!prev && curr) {
          // first subscriber joined
          this._start();
        }
        if (prev && !curr) {
          // last subscriber left
          this._stop();
        }
      });
  }

  attach(): Observable<ControllerState> {
    this._refCount$.next(this._refCount$.value + 1);
    return this._state$.asObservable().pipe(
      distinctUntilChanged(),
      finalize(() => this._refCount$.next(this._refCount$.value - 1))
    );
  }

  requestRefresh(): void {
    this._updates$.next();
  }

  _tearDown() {
    this._state$.next(ControllerState.UNKNOWN);
    this._lazy.unsubscribe();
    this._stop();
  }

  private _stop(): void {
    if (this._attach) {
      this._attach.unsubscribe();
      this._attach = undefined;
    }
  }

  private _start(): void {
    this._stop();
    this._attach = merge(
      merge(
        this._updates$.pipe(throttleTime(this._settings.chartControllerConfig().minRefresh)),
        this._notifications.messages(NotificationCategory.ActiveRecordingCreated),
        this._notifications.messages(NotificationCategory.ActiveRecordingDeleted),
        this._notifications.messages(NotificationCategory.ActiveRecordingStopped)
      ).pipe(switchMap((_) => this._target.target().pipe(first()))),
      this._target.target().pipe(tap((_) => this._state$.next(ControllerState.UNKNOWN)))
    )
      .pipe(concatMap((t) => this._hasRecording(t)))
      .subscribe((v) => {
        this._state$.next(v ? ControllerState.READY : ControllerState.NO_DATA);
        if (v) {
          this._api
            .uploadActiveRecordingToGrafana(RECORDING_NAME)
            .pipe(first())
            .subscribe((_) => {
              this._state$.next(ControllerState.READY);
            });
        }
      });
  }

  private _hasRecording(target: Target): Observable<boolean> {
    if (target === NO_TARGET) {
      return of(false);
    }

    return this._api
      .graphql<CountResponse>(
        `
          query ActiveRecordingsForAutomatedAnalysis($connectUrl: String) {
            targetNodes(filter: { name: $connectUrl }) {
              recordings {
                active (filter: {
                  labels: ["origin=${RECORDING_NAME}"],
                  state: "RUNNING",
                }) {
                aggregate {
                  count
                }
              }
            }
            }
          }`,
        { connectUrl: target.connectUrl }
      )
      .pipe(
        map((resp) => {
          const nodes = resp.data.targetNodes;
          if (nodes.length === 0) {
            return false;
          }
          const count = nodes[0].recordings.active.aggregate.count;
          return count > 0;
        }),
        catchError((_) => of(false))
      );
  }
}

interface CountResponse {
  data: {
    targetNodes: {
      recordings: {
        active: {
          aggregate: {
            count: number;
          };
        };
      };
    }[];
  };
}
