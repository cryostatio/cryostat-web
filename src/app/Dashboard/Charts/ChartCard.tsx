/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { CreateRecordingProps } from '@app/CreateRecording/CreateRecording';
import { LoadingView } from '@app/LoadingView/LoadingView';
import { ServiceContext } from '@app/Shared/Services/Services';
import { FeatureLevel } from '@app/Shared/Services/Settings.service';
import { useSubscriptions } from '@app/utils/useSubscriptions';
import {
  Bullseye,
  Button,
  CardActions,
  CardBody,
  CardHeader,
  CardTitle,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Label,
  Title,
} from '@patternfly/react-core';
import { DataSourceIcon, ExternalLinkAltIcon, SyncAltIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useHistory } from 'react-router-dom';
import { interval } from 'rxjs';
import { DashboardCardDescriptor, DashboardCardProps, DashboardCardSizes } from '../Dashboard';
import { DashboardCard } from '../DashboardCard';
import { ChartContext } from './ChartContext';
import { ControllerState, RECORDING_NAME } from './ChartController';

export interface ChartCardProps extends DashboardCardProps {
  theme: string;
  chartKind: string;
  duration: number;
  period: number;
}

// TODO these need to be localized
export enum ChartKind {
  'Core Count' = 1,
  'Thread Count' = 2,
  'CPU Load' = 3,
  'Heap Usage' = 4,
  'Memory Usage' = 5,
  'Total Memory' = 6,
  'Recording Start Time' = 7,
  'Recording Duration' = 8,
  'Classloading Statistics' = 9,
  'Metaspace Summary' = 10,
  'Network Utilization' = 11,
  'Metaspace GC Threshold' = 12,
  'Thread Statistics' = 13,
  'Exception Statistics' = 14,
  'Thread Context Switch Rate' = 15,
  'Compiler Statistics' = 16,

  'Safepoint Duration' = 18,
  'File I/O' = 19,
  'Compiler Total Time' = 20,

  'Compiler Peak Time' = 24,

  'Object Allocation Sample' = 38,
}

export function kindToId(kind: string): number {
  return ChartKind[kind];
}

export const ChartCard: React.FC<ChartCardProps> = (props) => {
  const [t] = useTranslation();
  const serviceContext = React.useContext(ServiceContext);
  const controllerContext = React.useContext(ChartContext);
  const history = useHistory();
  const addSubscription = useSubscriptions();
  const [controllerState, setControllerState] = React.useState(ControllerState.NO_DATA);
  const [randomKey, setRandomKey] = React.useState(Math.floor(Math.random()));
  const [chartSrc, setChartSrc] = React.useState('');
  const [dashboardUrl, setDashboardUrl] = React.useState('');

  const updateRandomKey = React.useCallback(() => {
    setRandomKey((prev) => {
      let next = prev + 1;
      if (next >= 10) {
        next = 0;
      }
      return next;
    });
  }, [setRandomKey]);

  React.useEffect(() => {
    addSubscription(serviceContext.api.grafanaDashboardUrl().subscribe(setDashboardUrl));
  }, [addSubscription, serviceContext, setDashboardUrl]);

  React.useEffect(() => {
    if (!dashboardUrl) {
      return;
    }
    const u = new URL('/d-solo/main', dashboardUrl);
    u.searchParams.append('theme', props.theme);
    u.searchParams.append('panelId', String(kindToId(props.chartKind)));
    u.searchParams.append('to', 'now');
    u.searchParams.append('from', `now-${props.duration}s`);
    u.searchParams.append('refresh', `${props.period}s`);
    setChartSrc(u.toString());
  }, [dashboardUrl, setControllerState, props.theme, props.chartKind, props.duration, props.period, setChartSrc]);

  React.useEffect(() => {
    addSubscription(controllerContext.controller.attach().subscribe(setControllerState));
  }, [addSubscription, controllerContext, setControllerState]);

  const refresh = React.useCallback(() => {
    controllerContext.controller.requestRefresh();
  }, [controllerContext]);

  React.useEffect(() => {
    refresh();
    addSubscription(interval(props.period * 1000).subscribe(() => refresh()));
  }, [addSubscription, props.period, refresh]);

  const popout = React.useCallback(() => {
    if (chartSrc && dashboardUrl) {
      window.open(chartSrc, '_blank');
    }
  }, [chartSrc, dashboardUrl]);

  const cardStyle = React.useMemo(() => {
    if (controllerState !== ControllerState.READY) {
      return {};
    }
    let height: number;
    switch (props.chartKind) {
      case 'Core Count':
        height = 250;
        break;
      default:
        height = 380;
        break;
    }
    return { height };
  }, [controllerState, props.chartKind]);

  const resyncButton = React.useMemo(() => {
    return (
      <Button
        key={0}
        aria-label={t('CHART_CARD.BUTTONS.SYNC.LABEL', { chartKind: props.chartKind })}
        onClick={updateRandomKey}
        variant="plain"
        icon={<SyncAltIcon />}
        isDisabled={!chartSrc || !dashboardUrl}
      />
    );
  }, [t, props.chartKind, updateRandomKey, chartSrc, dashboardUrl]);

  const popoutButton = React.useMemo(() => {
    return (
      <Button
        key={1}
        aria-label={t('CHART_CARD.BUTTONS.POPOUT.LABEL', { chartKind: props.chartKind })}
        onClick={popout}
        variant="plain"
        icon={<ExternalLinkAltIcon />}
        isDisabled={!chartSrc || !dashboardUrl}
      />
    );
  }, [t, props.chartKind, popout, chartSrc, dashboardUrl]);

  const actions = React.useMemo(() => {
    const a = props.actions || [];
    return [resyncButton, popoutButton, ...a];
  }, [props.actions, resyncButton, popoutButton]);

  const header = React.useMemo(() => {
    const isWide = props.span > 4;
    const style = {
      marginBottom: isWide ? '-2em' : '',
    };
    if (controllerState === ControllerState.READY) {
      return (
        <CardHeader style={style}>
          <CardActions>{actions}</CardActions>
        </CardHeader>
      );
    } else {
      return (
        <CardHeader style={style}>
          <CardTitle>{props.chartKind}</CardTitle>
          <CardActions>{props.actions}</CardActions>
        </CardHeader>
      );
    }
  }, [props.actions, props.span, props.chartKind, controllerState, actions]);

  const handleCreateRecording = React.useCallback(() => {
    history.push({
      pathname: '/recordings/create',
      state: {
        restartExisting: true,
        name: RECORDING_NAME,
        templateName: 'Continuous',
        templateType: 'TARGET',
        labels: [{ key: 'origin', value: RECORDING_NAME }],
        duration: -1,
        // TODO these are arbitrary defaults that will be set in the recording creation form.
        // Should these values be inferred in some more intelligent way?
        maxAge: 120, // seconds
        maxSize: 100 * 1024 * 1024, // bytes
      } as CreateRecordingProps,
    });
  }, [history]);

  return (
    <DashboardCard
      id={props.chartKind + '-chart-card'}
      dashboardId={props.dashboardId}
      cardSizes={ChartCardSizes}
      isCompact
      style={cardStyle}
      cardHeader={header}
    >
      <CardBody>
        {controllerState === ControllerState.UNKNOWN ? (
          <LoadingView />
        ) : controllerState === ControllerState.READY ? (
          <iframe
            className="disabled-pointer"
            key={controllerState + randomKey}
            style={{ height: '100%', width: '100%' }}
            src={chartSrc}
          />
        ) : (
          <Bullseye>
            <EmptyState variant={EmptyStateVariant.large}>
              <EmptyStateIcon icon={DataSourceIcon} />
              <Title headingLevel="h2" size="md">
                {t('CHART_CARD.NO_RECORDING.TITLE')}
              </Title>
              <EmptyStateBody>
                <Trans
                  t={t}
                  values={{ recordingName: RECORDING_NAME }}
                  components={{ label: <Label color="blue" isCompact /> }}
                >
                  CHART_CARD.NO_RECORDING.DESCRIPTION
                </Trans>
              </EmptyStateBody>
              <Button variant="primary" onClick={handleCreateRecording}>
                {t('CHART_CARD.BUTTONS.CREATE.LABEL')}
              </Button>
            </EmptyState>
          </Bullseye>
        )}
      </CardBody>
    </DashboardCard>
  );
};

export const ChartCardSizes: DashboardCardSizes = {
  span: {
    minimum: 3,
    default: 4,
    maximum: 12,
  },
  height: {
    // TODO: implement height resizing
    minimum: Number.NaN,
    default: Number.NaN,
    maximum: Number.NaN,
  },
};

export const ChartCardDescriptor: DashboardCardDescriptor = {
  featureLevel: FeatureLevel.BETA,
  title: 'CHART_CARD.CARD_TITLE',
  cardSizes: ChartCardSizes,
  description: 'CHART_CARD.CARD_DESCRIPTION',
  descriptionFull: 'CHART_CARD.CARD_DESCRIPTION_FULL',
  component: ChartCard,
  propControls: [
    {
      name: 'CHART_CARD.PROP_CONTROLS.THEME.NAME',
      key: 'theme',
      values: ['light', 'dark'],
      defaultValue: 'light',
      description: 'CHART_CARD.PROP_CONTROLS.THEME.DESCRIPTION',
      kind: 'select',
    },
    {
      name: 'CHART_CARD.PROP_CONTROLS.PERFORMANCE_METRIC.NAME',
      key: 'chartKind',
      values: Object.values(ChartKind).filter((v) => typeof v === 'string'),
      defaultValue: Object.values(ChartKind).filter((v) => typeof v === 'string')[0],
      description: 'CHART_CARD.PROP_CONTROLS.PERFORMANCE_METRIC.DESCRIPTION', // TODO should this be a function that returns a value based on the selection?
      kind: 'select',
    },
    {
      name: 'CHART_CARD.PROP_CONTROLS.DATA_WINDOW.NAME',
      key: 'duration',
      defaultValue: 120,
      description: 'CHART_CARD.PROP_CONTROLS.DATA_WINDOW.DESCRIPTION',
      kind: 'number',
      extras: {
        min: 30,
        max: 300,
      },
    },
    {
      name: 'CHART_CARD.PROP_CONTROLS.REFRESH_PERIOD.NAME',
      key: 'period',
      defaultValue: 10,
      description: 'CHART_CARD.PROP_CONTROLS.REFRESH_PERIOD.DESCRIPTION',
      kind: 'number',
      extras: {
        min: 5,
        max: 120,
      },
    },
  ],
};
