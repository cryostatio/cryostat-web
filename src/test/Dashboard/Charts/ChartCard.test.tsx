/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

jest.mock('@app/Dashboard/Charts/ChartController');

import { ChartCard, kindToId } from '@app/Dashboard/Charts/ChartCard';
import { ChartContext } from '@app/Dashboard/Charts/ChartContext';
import { ChartController, ControllerState } from '@app/Dashboard/Charts/ChartController';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';
import { setupStore, store } from '@app/Shared/Redux/ReduxStore';
import { defaultServices, ServiceContext } from '@app/Shared/Services/Services';
import { cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory, MemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import renderer, { act } from 'react-test-renderer';
import { of } from 'rxjs';
import { renderWithProvidersAndRedux } from '../../Common';

let history: MemoryHistory = createMemoryHistory({ initialEntries: ['/'] });
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: () => ({ url: history.location.pathname }),
  useHistory: () => history,
}));

const mockDashboardUrl = 'http://localhost:3000';
jest.spyOn(defaultServices.api, 'grafanaDashboardUrl').mockReturnValue(of(mockDashboardUrl));

const mockTarget = { connectUrl: 'service:jmx:rmi://someUrl', alias: 'fooTarget' };
jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));

const mockController = new ChartController(
  defaultServices.api,
  defaultServices.target,
  defaultServices.notificationChannel,
  defaultServices.settings
);
const mockChartContext = {
  controller: mockController,
};

describe('<ChartCard />', () => {
  beforeEach(() => (history = createMemoryHistory({ initialEntries: ['/'] })));
  afterEach(cleanup);

  it('renders correctly', async () => {
    jest.spyOn(mockController, 'attach').mockReturnValue(of(ControllerState.READY));

    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <ChartContext.Provider value={mockChartContext}>
              <Provider store={store}>
                <ChartCard theme={'light'} chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />
              </Provider>
            </ChartContext.Provider>
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot('with-content');
  });

  it('renders loading state correctly', async () => {
    jest.spyOn(mockController, 'attach').mockReturnValue(of(ControllerState.UNKNOWN));

    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <ChartContext.Provider value={mockChartContext}>
              <Provider store={store}>
                <ChartCard theme={'light'} chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />
              </Provider>
            </ChartContext.Provider>
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot('loading-view');
  });

  it('renders empty state correctly', async () => {
    jest.spyOn(mockController, 'attach').mockReturnValue(of(ControllerState.NO_DATA));

    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <ChartContext.Provider value={mockChartContext}>
              <Provider store={store}>
                <ChartCard theme={'light'} chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />
              </Provider>
            </ChartContext.Provider>
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot('empty-state');
  });

  it('renders empty state with information and action button', async () => {
    jest.spyOn(mockController, 'attach').mockReturnValue(of(ControllerState.NO_DATA));

    renderChartCard(
      <ChartCard theme={'light'} chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />
    );

    expect(screen.getByText('CPU Load')).toBeInTheDocument();
    expect(screen.getByText('No source recording')).toBeInTheDocument();
    expect(
      screen.getByText((s) => s.includes('Metrics cards display data taken from running flight recordings'))
    ).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /create/i })).toBeInTheDocument();
  });

  it('navigates to recording creation with prefilled state when empty state button clicked', async () => {
    jest.spyOn(mockController, 'attach').mockReturnValue(of(ControllerState.NO_DATA));

    const { user } = renderChartCard(
      <ChartCard theme={'light'} chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />
    );

    expect(history.location.pathname).toBe('/');
    await user.click(screen.getByRole('button', { name: /create/i }));
    expect(history.location.pathname).toBe('/recordings/create');
    expect(history.location.state).toEqual({
      duration: -1,
      labels: [
        {
          key: 'origin',
          value: 'dashboard_metrics',
        },
      ],
      maxAge: 120,
      maxSize: 100 * 1024 * 1024,
      name: 'dashboard_metrics',
      restartExisting: true,
      templateName: 'Continuous',
      templateType: 'TARGET',
    });
  });

  it.each([
    ['CPU Load', 120, 5],
    ['Heap Usage', 90, 10],
    ['Network Utilization', 60, 15],
    ['File I/O', 30, 20],
  ])('renders iframe', async (chartKind: string, duration: number, period: number) => {
    jest.spyOn(mockController, 'attach').mockReturnValue(of(ControllerState.READY));

    const { container } = renderChartCard(
      <ChartCard theme={'light'} chartKind={chartKind} duration={duration} period={period} span={6} dashboardId={0} />
    );

    const iframe = container.querySelector('iframe');
    expect(iframe).toBeTruthy();
    const u = new URL(iframe?.src || '');
    expect(u.host).toBe('localhost:3000');
    expect(u.protocol).toBe('http:');
    const params = new URLSearchParams();
    params.set('theme', 'light');
    params.set('panelId', String(kindToId(chartKind)));
    params.set('to', 'now');
    params.set('from', `now-${duration}s`);
    params.set('refresh', `${period}s`);
    expect(u.searchParams.toString()).toEqual(params.toString());
  });
});

const renderChartCard = (
  ui: React.ReactElement,
  {
    services = defaultServices,
    notifications = NotificationsInstance,
    chartContext = mockChartContext,
    preloadState = {},
    store = setupStore(preloadState),
    user = userEvent.setup(),
    ...renderOptions
  } = {}
) => {
  return renderWithProvidersAndRedux(
    ui,
    [
      {
        kind: ChartContext.Provider,
        instance: chartContext,
      },
      {
        kind: NotificationsContext.Provider,
        instance: notifications,
      },
      {
        kind: ServiceContext.Provider,
        instance: services,
      },
    ],
    {
      store,
      user,
      ...renderOptions,
    }
  );
};
