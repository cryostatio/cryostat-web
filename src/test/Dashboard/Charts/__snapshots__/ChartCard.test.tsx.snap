// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartCard /> renders correctly 1`] = `
<div
  className="draggable-ref-wrapper"
  onDragStart={[Function]}
  onTransitionEnd={[Function]}
  style={Object {}}
>
  <div
    className="dashboard-card-resizable-wrapper"
  >
    <article
      className="pf-c-card pf-m-compact pf-m-rounded dashboard-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF4/Card"
      data-ouia-safe={true}
      id="CPU Load-chart-card"
      style={
        Object {
          "height": 380,
        }
      }
    >
      <div
        className="draggable-ref__grip"
        draggable={true}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="pf-c-card__header"
          style={
            Object {
              "marginBottom": "-2.8em",
            }
          }
        >
          <div
            className="pf-c-card__actions"
          >
            <button
              aria-disabled={false}
              aria-label="Expand CPU Load chart window"
              className="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M370.72 133.28C339.458 104.008 298.888 87.962 255.848 88c-77.458.068-144.328 53.178-162.791 126.85-1.344 5.363-6.122 9.15-11.651 9.15H24.103c-7.498 0-13.194-6.807-11.807-14.176C33.933 94.924 134.813 8 256 8c66.448 0 126.791 26.136 171.315 68.685L463.03 40.97C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.749zM32 296h134.059c21.382 0 32.09 25.851 16.971 40.971l-41.75 41.75c31.262 29.273 71.835 45.319 114.876 45.28 77.418-.07 144.315-53.144 162.787-126.849 1.344-5.363 6.122-9.15 11.651-9.15h57.304c7.498 0 13.194 6.807 11.807 14.176C478.067 417.076 377.187 504 256 504c-66.448 0-126.791-26.136-171.315-68.685L48.97 471.03C33.851 486.149 8 475.441 8 454.059V320c0-13.255 10.745-24 24-24z"
                />
              </svg>
            </button>
            <button
              aria-disabled={false}
              aria-label="Pop out CPU Load chart"
              className="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="pf-c-card__body"
      >
        <iframe
          src="http://localhost:3000/d-solo/main?theme=light&panelId=3&to=now&from=now-120s&refresh=10s"
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </div>
    </article>
    <div
      className="resizable-ref"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;
